### ISSUE TITLE
stack setup: "expected minor version match with ghc-3.10.7"

### ISSUE THREAD

**Issue by nd2s** [commercialhaskell/stack#1903]
Not sure if I'm doing something wrong or if it's a bug. With `lts-5.1` - `lts-5.6` I'm getting following error messages. First on `stack install`:

```
Fetched package index.    
Populated index cache.    
Compiler version mismatched, found ghc-7.10.2 (x86_64), but expected minor version match with ghc-7.10.3 (x86_64) (based on resolver setting in /usr/home/.../stack.yaml).
Try running "stack setup" to install the correct GHC into /usr/home/.../.stack/programs/x86_64-freebsd/
```

This is OK. But then then on `stack setup`:
```
No information found for ghc-3.10.7.
Supported versions for OS key 'freebsd64': GhcVersion 7.8.4, GhcVersion 7.10.1, GhcVersion 7.10.2, GhcVersion 7.10.3
```

Why `3.10.7`? It seems the version number is messed up here?

Stack `Version 1.0.4.2 x86_64` on `FreeBSD 10.2 amd64`.

**Comment by mgsloan** [comment]
Uhh, something must be really screwed up with version numbers because as far as I know, `1.0.4.2` is not a version of stack that has ever existed.

Can you please try to install stack 1.0.4 from tarball?

**Comment by nd2s** [comment]
I think you are wrong. I just installed that version with `cabal install`. It's on hackage: https://hackage.haskell.org/package/stack . Will try with `1.0.4` as well.

**Comment by mgsloan** [comment]
Oh, I see.  I didn't see such a version here https://github.com/commercialhaskell/stack/releases , they must have been cabal metadata updates.

**Comment by nd2s** [comment]
I tried with `1.0.4`: There `stack install` also shows me wrong version number `3.10.7` (instead of `7.10.3`).

**Comment by centromere** [comment]
I too am getting the same error on FreeBSD. Additionally, I get this error:

`The following package identifiers were not found in your indices: aeson-1.0.9.0`.

**Comment by mgsloan** [comment]
Both you and @centromere built stack using `cabal-install`.  I'm guessing that there's some dependency that it's choosing which is causing some funky behavior, which is why other users aren't seeing the problem.

**Comment by nd2s** [comment]
Well, what are the alternatives? I can't use stack to build stack, that I need for building things with stack. :) Unfortunately there is no stack port on FreeBSD.

I will try building master branch tomorrow.

**Comment by mgsloan** [comment]
You could try building it in a cabal sandbox so it's less affected by system state.  Also good to make sure there isn't superfluous stuff in your global DB (just the things that come with GHC).  @centromere sent me a `cabal.config`, but unfortunately I still cannot reproduce the issue:

```
Version 1.0.4.2, Git revision d57f56e60ff7e613bdb7364a9a345159bcf8653a (dirty) (3231 commits) x86_64
mgsloan@computer:~/fpco/stack/.cabal-sandbox/bin$ ./stack setup
Downloaded lts-5.3 build plan.    
Caching build plan
Fetched package index.                                                                                    
Populated index cache.    
Fetched package index.                                                                                    
Populated index cache.    
Preparing to install GHC to an isolated location.
This will not interfere with any system-level installation.
Downloaded ghc-7.10.3.                                     
Installing GHC ...  
```

I'm not that familiar with cabal sandboxes (relied on hsenv back in the pre-stack days).  The sandbox will pay attention to `cabal.config`, right?

So, this doesn't seem to be an issue with dependencies, but something somehow FreeBSD specific.  I gotta say, this is probably the most bizarre issue in the history of stack XD

**Comment by nd2s** [comment]
I don't know about `cabal.config`, sorry.

I also tried building in a sandbox with the latest package from Hackage. Makes no difference. I haven't had any other packages installed before compiling stack anyway. Except for some Haskell ports that were installed globally (dependencies of xmonad and such).

I ignored another error message of `cabal install/setup` output yesterday. That was a mistake, this seems  to be important:
```
stack will almost certainly fail with GHC below version 7.8
Valiantly attempting to run anyway, but I know this is doomed
For more information, see: https://github.com/commercialhaskell/stack/issues/648
``` 
There is a problem as well, because: `The Glorious Glasgow Haskell Compilation System, version 7.10.2`.

How are version strings parsed/generated? Is it possible that someone (in some dependency) is reversing them with `foldl` or something like that?

**Comment by macalimlim** [comment]
if you are using archlinux, just install the stack that came with arch and use that (might be possible with other distros too) for the moment...

**Comment by nd2s** [comment]
@macalimlim: This ticket has nothing to do with Linux.

**Comment by macalimlim** [comment]
Yes, what I'm saying is use the stack that came from your linux distro just to avoid that error.

**Comment by nd2s** [comment]
@macalimlim: And what I'm saying is there is no Linux distribution. Please read the ticket.

**Comment by macalimlim** [comment]
Ok sorry for the confusion. Im just making a note in case a user from a linux distro encounters the same error.

**Comment by mircea-cosbuc** [comment]
@macalimlim Actually upgrading stack from 1.0.4.2-1 to 1.0.4.2-2 on archlinux introduces this issue

**Comment by macalimlim** [comment]
Actually when I installed my stack, its still on 1.0.4.2-1 upon encountering this error. If upgrading it to 1.0.4.2-2 introduces this issue then just revert it back to 1.0.4.2-1. At this point in time I still don't see the update from our mirror server though... Anyways thanks for pointing it out ;)

**Comment by mkoloberdin** [comment]
I can confirm that downgrading to 1.0.4.2-1 on Arch Linux resolves the issue.

The commit message for 1.0.4.2-2 says:

"rebuild with conduit-1.2.6.4, http2-1.5.2, scientific-0.3.4.6, vector-binary-instances-0.2.3.0"

So this bug must have something to do with changes in one of those packages.

**Comment by macalimlim** [comment]
Seems bumping lts-5.7 to lts-5.8 resolved this. Can anyone confirm?
Based on the changelog (from lts-5.7 to lts-5.8), vector-binary-instances got upgraded confirming the changes that were made to stack 1.0.4.2-2 as pointed out by @mkoloberdin. On my machine I rebuilt stack (from stackage) with stack (from distro, version 1.0.4.2-1) and used stack from stackage...

**Comment by mkoloberdin** [comment]
@macalimlim,
Looks like you a right, the culprit is vector-binary-instances-0.2.3.0:
https://github.com/bos/vector-binary-instances/commit/52e50365a07eb598190cc2ed2a447e5ac6cc449e
"Fix reversal of decoded output"

**Comment by mgsloan** [comment]
Yup, here's the associated issue https://github.com/bos/vector-binary-instances/issues/6

**Comment by felixonmars** [comment]
The Arch package should have been fixed in 1.0.4.2-3.

**Comment by nd2s** [comment]
I can confirm that the problem is solved. Ticket can be closed.
